src/hotspot/share/code/codeCache.cpp

```diff
@@ -178,143 +178,144 @@ GrowableArray<CodeHeap*>* CodeCache::_compiled_heaps = new(mtCode) GrowableArray
-        // Adapt size of profiled code heap
-        if (diff_size < 0 && ((intx)profiled_size + diff_size) <= 0) {
-          // Not enough space available, set to minimum size
-          diff_size += profiled_size - min_size;
-          profiled_size = min_size;
-        } else {
-          profiled_size += diff_size;
-          diff_size = 0;
-        }
-      }
-    } else if (profiled_set) {
-      // Adapt size of non-profiled code heap
-      if (diff_size < 0 && ((intx)non_profiled_size + diff_size) <= 0) {
-        // Not enough space available, set to minimum size
-        diff_size += non_profiled_size - min_size;
-        non_profiled_size = min_size;
-      } else {
-        non_profiled_size += diff_size;
-        diff_size = 0;
-      }
-    } else if (non_nmethod_set) {
-      // Distribute remaining size between profiled and non-profiled code heaps
-      diff_size = cache_size - non_nmethod_size;
-      profiled_size = diff_size / 2;
-      non_profiled_size = diff_size - profiled_size;
-      diff_size = 0;
-    }
-    if (diff_size != 0) {
-      // Use non-nmethod code heap for remaining space requirements
-      assert(!non_nmethod_set && ((intx)non_nmethod_size + diff_size) > 0, "sanity");
-      non_nmethod_size += diff_size;
-    }
-  }
+  CodeHeapInfo non_nmethod = {NonNMethodCodeHeapSize, FLAG_IS_CMDLINE(NonNMethodCodeHeapSize), true};
+  CodeHeapInfo profiled = {ProfiledCodeHeapSize, FLAG_IS_CMDLINE(ProfiledCodeHeapSize), true};
+  CodeHeapInfo non_profiled = {NonProfiledCodeHeapSize, FLAG_IS_CMDLINE(NonProfiledCodeHeapSize), true};
 
-  // We do not need the profiled CodeHeap, use all space for the non-profiled CodeHeap
+  const bool cache_size_set   = FLAG_IS_CMDLINE(ReservedCodeCacheSize);
+  const size_t ps             = page_size(false, 8);
+  const size_t min_size       = MAX2(os::vm_allocation_granularity(), ps);
+  const size_t min_cache_size = CodeCacheMinimumUseSpace DEBUG_ONLY(* 3); // Make sure we have enough space for VM internal code
+  size_t cache_size           = ReservedCodeCacheSize;
+
+  // Prerequisites
   if (!heap_available(CodeBlobType::MethodProfiled)) {
-    non_profiled_size += profiled_size;
-    profiled_size = 0;
+    // For compatibility reason, disabled tiered compilation overrides
+    // segment size ever if it was set explicitly.
+    non_profiled.size += profiled.size;
+    // Profiled code heap is not available, forcibly set size to 0
+    profiled.size = 0;
+    profiled.set = true;
+    profiled.enabled = false;
   }
-  // We do not need the non-profiled CodeHeap, use all space for the non-nmethod CodeHeap
+
   if (!heap_available(CodeBlobType::MethodNonProfiled)) {
-    non_nmethod_size += non_profiled_size;
-    non_profiled_size = 0;
+    // Compatibility
+    non_nmethod.size += profiled.size;
+    // Non-Profiled code heap is not available, forcibly set size to 0
+    non_profiled.size = 0;
+    non_profiled.set = true;
+    non_profiled.enabled = false;
+  }
```

Comment:
It can be rewritten to the shorter version:
```
// If either heap1 or heap2 is not available, its size is added to the size of the available heap.
static void reuse_unavailable_heap_size(CodeHeapInfo *heap1, CodeHeapInfo *heap2) {
   assert(CodeCache::heap_available(heap1->type) || CodeCache::heap_available(heap2->type), "At least one code heap must be available");
   if (!CodeCache::heap_available(heap1->type)) {
     swap(heap1, heap2);
   } else if (CodeCache::heap_available(heap2->type)) {
     // Both code heaps are available. Nothing needs to be done.
     return;
   }
   heap1->size += heap2->size;
   heap2->size = 0;
   heap2->set = true;
   heap2->enable = false;
}
```

Now we can replace those two IFs with:
```
// For compatibility we add the size of an unavailable code heap to the size of the available heap.
reuse_unavailable_heap_size(&non_profiled, &profiled);
```

Reply:
That sounds reasonable. But that is kind of the opposite of what we are doing with this change. We are moving to a plain, simple, readable code.